src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -5855,6 +5867,208 @@ private BoundExpression BindTypeParameterCreationExpression(ObjectCreationExpres
             return MakeBadExpressionForObjectCreation(node, typeParameter, analyzedArguments, initializerOpt, typeSyntax, diagnostics);
         }
 
+        private BoundExpression BindCollectionLiteralExpression(CollectionCreationExpressionSyntax syntax, TypeSymbol targetType, BindingDiagnosticBag diagnostics)
+        {
+            // Follows approach from BindCollectionInitializerExpression.
+
+            // PROTOTYPE: Test types other than NamedTypeSymbols such as arrays and type parameters.
+            // PROTOTYPE: Test constructor with all optional parameters; with params parameter.
+            var constructor = ((NamedTypeSymbol)targetType).Constructors.SingleOrDefault(c => c.ParameterCount == 0);
+            if (constructor is null)
+            {
+                // PROTOTYPE: Report error.
+                return BadExpression(syntax);
+            }
+
+            bool hasEnumerableInitializerType = CollectionInitializerTypeImplementsIEnumerable(targetType, syntax, diagnostics);
+            if (!hasEnumerableInitializerType && !syntax.HasErrors && !targetType.IsErrorType())
+            {
+                Error(diagnostics, ErrorCode.ERR_CollectionLiteralTargetTypeNotConstructible, syntax, targetType);
+            }
+
+            var implicitReceiver = new BoundObjectOrCollectionValuePlaceholder(syntax, isNewInstance: true, targetType) { WasCompilerGenerated = true };
+            var collectionInitializerAddMethodBinder = this.WithAdditionalFlags(BinderFlags.CollectionInitializerAddMethod);
+
+            var initializerBuilder = ArrayBuilder<BoundExpression>.GetInstance();
+            foreach (var elementInitializer in syntax.Elements)
+            {
+                initializerBuilder.Add(
+                    BindCollectionLiteralElement(elementInitializer, hasEnumerableInitializerType, collectionInitializerAddMethodBinder, diagnostics, implicitReceiver));
+            }
+
+            return new BoundCollectionLiteralExpression(
+                syntax,
+                constructor,
+                implicitReceiver,
+                initializerBuilder.ToImmutableAndFree(),
+                targetType);
+        }
+
+        private BoundExpression BindCollectionLiteralElement(
+            CollectionElementSyntax syntax,
+            bool hasEnumerableInitializerType,
+            Binder collectionInitializerAddMethodBinder,
+            BindingDiagnosticBag diagnostics,
+            BoundObjectOrCollectionValuePlaceholder implicitReceiver)
+        {
+            switch (syntax)
+            {
+                case ExpressionElementSyntax expressionElementSyntax:
+                    {
+                        var elementSyntax = expressionElementSyntax.Expression;
+                        var element = BindValue(elementSyntax, diagnostics, BindValueKind.RValue);
```

Comment:
this likely needs a prototype comment right?  or can you explain how this works with, say, binding `null` here?  What brings the target type of hte collection-element type into this BindValue?

Reply:
tbh, i'm a little confused by this.  shouldn't this be bound as if it were being used in `receiver.Add(expr)`?

Reply:
We should consider using indexer assignment rather than `Add()` _always_ (or _never_) when the target type is a dictionary, so the semantics are always _replace_ (or _add_) for both `k:v` and `e` elements.